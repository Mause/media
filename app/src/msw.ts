import { setupServer } from 'msw/node';

export const server = setupServer();
server.events.on('request:start', ({ request }) => {
  console.log('Request start:', request.method, request.url);
});
server.events.on('request:match', ({ request }) => {
  console.log('Request match:', request.method, request.url);
});
server.events.on('request:end', ({ request }) => {
  console.log('Request end:', request.method, request.url);
});
server.events.on('request:unhandled', ({ request }) => {
  console.error('Request unhandled:', request.method, request.url);
});
