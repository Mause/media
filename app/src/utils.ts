import { String } from 'typescript-string-operations';
import { useState, useEffect } from 'react';
import type { RawAxiosRequestHeaders } from 'axios';
import Axios from 'axios';
import * as RRD from 'react-router-dom';
// import axiosRetry from '@vtex/axios-concurrent-retry';
import type {
  Auth0ContextInterface,
  AuthenticationError,
} from '@auth0/auth0-react';
import { useAuth0 } from '@auth0/auth0-react';
import moment from 'moment';
import * as _ from 'lodash-es';

import { FetchEventTarget } from './fetch_stream';
import type { TV } from './SeasonSelectComponent';
import type { EpisodeResponse } from './streaming';

// axiosRetry(Axios, { retries: 3 });

export type GetResponse<T> = T extends {
  get: { responses: { '200': { content: { 'application/json': unknown } } } };
}
  ? T['get']['responses']['200']['content']['application/json']
  : never;

export function subscribe<T>(
  path: string,
  callback: (a: T) => void,
  error: (e: Error) => void,
  authorization: string,
  end: (() => void) | null = () => null,
): () => void {
  const es = FetchEventTarget(path, {
    headers: new Headers({
      Authorization: `Bearer ${authorization}`,
    }),
  });
  const onerror = (event: Event) => {
    error(event as unknown as Error);
  };
  es.addEventListener('abort', onerror);
  const internal_callback = (event: Event) => {
    callback((event as MessageEvent).data as T);
  };
  es.addEventListener('message', internal_callback);

  return () => {
    es.removeEventListener('close', end);
    es.removeEventListener('abort', onerror);
    es.removeEventListener('message', internal_callback);
  };
}

export function getPrefix() {
  const prefix = import.meta.env.REACT_APP_API_PREFIX;

  if (!prefix) {
    return '';
  } else if (prefix.includes('localhost')) {
    return '';
  } else if (prefix) {
    return `https://${prefix}`;
  }
}

export async function load<T>(
  path: string,
  params?: object,
  headers?: RawAxiosRequestHeaders,
): Promise<T> {
  const url = `${getPrefix()}/api/${path}`;
  console.log('Loading', { url, params });
  const t = await Axios.get<T>(url, {
    params,
    withCredentials: true,
    headers,
  });
  return t?.data;
}

interface Res<T> {
  data?: T;
  done: boolean;
  error?: Error;
}

function isAuthenticationError(e: unknown): e is AuthenticationError {
  return typeof e === 'object' && e != null && 'error' in e;
}

export async function getToken(auth0: Auth0ContextInterface): Promise<string> {
  try {
    return await auth0.getAccessTokenSilently();
  } catch (e) {
    if (isAuthenticationError(e) && e.error === 'missing_refresh_token') {
      await auth0.loginWithRedirect({
        authorizationParams: {
          redirect_uri: window.location.toString(),
        },
      });
      return '';
    } else {
      throw e;
    }
  }
}

export function usePost<T>(
  url: string,
  body: object,
): { done: boolean; data?: T; error?: Error } {
  const [result, setResult] = useState<Res<T>>({ done: false });
  const auth = useAuth0();

  useEffect(() => {
    const abortController = new AbortController();
    getToken(auth)
      .then((token) =>
        Axios.post<T>(`/api/${url}`, body, {
          signal: abortController.signal,
          headers: {
            Authorization: `Bearer ${token}`,
          },
        }),
      )
      .then(
        (res) => {
          setResult({ done: true, data: res.data });
        },
        (error: unknown) => {
          setResult({ done: true, error: error as Error });
        },
      );

    return () => {
      abortController.abort();
    };
  }, [url, body, auth]);

  return result;
}

export function useLocation<T>() {
  const location = RRD.useLocation();
  return { ...location, state: location.state as T };
}

export function getMarker(episode: {
  season?: number;
  episode?: number | null;
}) {
  if (episode.episode) {
    return String.format('S{0:00}E{1:00}', episode.season, episode.episode);
  } else {
    return String.format('S{0:00}', episode.season);
  }
}

export function getMessage(air_date: string) {
  const today = moment().startOf('day');
  const tomorrow = today.clone().add(1, 'day');
  const yesterday = today.clone().subtract(1, 'day');
  const dt = moment(air_date);
  const dts = dt.format('DD/MM/YYYY');

  let message;
  if (today.isSame(dt)) {
    message = 'airs today';
  } else if (dt.isSame(yesterday)) {
    message = 'aired yesterday';
  } else if (dt.isSame(tomorrow)) {
    message = 'airs tomorrow';
  } else if (dt.isAfter(today)) {
    message = `airs on ${dts}`;
  } else {
    message = `aired on ${dts}`;
  }
  return message;
}

export function shouldCollapse(
  i: string,
  data: TV | undefined,
  episodes: EpisodeResponse[],
): boolean {
  let collapse = false;
  if (data) {
    const i_i = +i;
    const seasonMeta = data.seasons.find((s) => s.season_number === i_i);
    if (seasonMeta) {
      const hasNext = true; // !!data.seasons[i_i + 1];

      const episodeNumbers = _.range(1, seasonMeta.episode_count + 1);
      const hasNumbers = _.map(episodes, 'episode');
      const hasAllEpisodes =
        _.difference(episodeNumbers, hasNumbers).length === 0;
      collapse = hasNext && hasAllEpisodes;
    }
  }
  return collapse;
}
