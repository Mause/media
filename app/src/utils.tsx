import React, { ReactElement } from 'react';
import Axios from 'axios';
import { useState, useEffect } from 'react';
import MaterialLink from '@material-ui/core/Link';
import { Link } from 'react-router-dom';
import { LocationDescriptor } from 'history';
import axiosRetry from '@vtex/axios-concurrent-retry';
import { TypographyTypeMap } from '@material-ui/core';
import moxios from 'moxios';
import { unstable_batchedUpdates } from 'react-dom';

// axiosRetry(Axios, { retries: 3 });

export function MLink<S>(props: {
  children: React.ReactNode;
  to: LocationDescriptor<S>;
  color?: TypographyTypeMap['props']['color'];
}): ReactElement {
  return <MaterialLink component={Link} {...props} />;
}

export function subscribe<T>(
  path: string,
  callback: (a: T) => void,
  end: (() => void) | null = null,
): void {
  const es = new EventSource(path, {
    withCredentials: true,
  });
  es.addEventListener('message', ({ data }) => {
    if (!data) {
      if (end) {
        end();
      }
      return es.close();
    }
    callback(JSON.parse(data));
  });
}

export async function load<T>(path: string, params?: string): Promise<T> {
  const t = await Axios.get<T>(`/api/${path}`, {
    params,
    withCredentials: true,
  });
  return t && t.data;
}

export function usePost<T>(
  url: string,
  body: object,
): { done: boolean; data?: T; error?: Error } {
  const [done, setDone] = useState<boolean>(false);
  const [error, setError] = useState<Error>();
  const [data, setData] = useState<T>();

  useEffect(
    () => {
      Axios.post<T>('/api/' + url, body, {
        withCredentials: true,
      }).then(
        res => {
          unstable_batchedUpdates(() => {
            setDone(true);
            setData(res.data);
          });
        },
        error => {
          unstable_batchedUpdates(() => {
            setDone(true);
            setError(error);
          });
        },
      );
    },
    [url, body],
  );

  return { done, error, data };
}

export function ExtMLink(props: { href: string; children: string }) {
  return (
    <MaterialLink href={props.href} target="_blank" rel="noopener noreferrer">
      {props.children}
    </MaterialLink>
  );
}

export function expectLastRequestBody() {
  return expect(JSON.parse(moxios.requests.mostRecent().config.data));
}
