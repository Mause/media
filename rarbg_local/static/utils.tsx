import Axios from 'axios';
import { useState, useEffect } from 'react';

export function subscribe(path: string, callback: (a: any) => void, end: (() => void) | null = null): void {
  const es = new EventSource(path, {
    withCredentials: true,
  });
  es.addEventListener('message', ({ data }) => {
    if (!data) {
      if (end) {
        end();
      }
      return es.close();
    }
    callback(JSON.parse(data));
  });
}

export function load<T>(path: string, params?: any): Promise<T> {
  return Axios.get<T>(`/api/${path}`, { params, withCredentials: true }).then(t => t.data);
}

export function useLoad<T>(path: string, params: any = null) {
  const [data, setData] = useState<T>()
  useEffect(() => {
    load<T>(path, params).then(setData);
  }, [path, params]);
  return data;
}
