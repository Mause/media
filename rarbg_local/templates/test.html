{% extends "base.html" %}

{% block header %}
Options for "{{title}}"
{% endblock %}

{% block content %}
<script type="text/javascript">
  var require = {
    bundles: JSON.parse(`{{BUNDLES|safe}}`),
    deps: ['streaming'],
    callback: function () {
      console.log('Loaded');
    }
  };
</script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.js"></script>

<div id="app"></div>
{% endblock %}
