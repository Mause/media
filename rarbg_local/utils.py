from functools import lru_cache as _lru_cache
from typing import Set

_caches: Set[_lru_cache] = set()  # type: ignore


def lru_cache(*args, **kwargs):
    def wrapper(func):
        cache = _lru_cache(*args, **kwargs)(func)
        _caches.add(cache)
        return cache

    return wrapper


def cache_clear():
    while _caches:
        cache = _caches.pop()
        cache.cache_clear()  # type: ignore
