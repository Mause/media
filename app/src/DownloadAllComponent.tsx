import { useParams } from 'react-router-dom';
import useSWR from 'swr';
import { ITorrent, DisplayTorrent } from './OptionsComponent';
import React from 'react';
import { Loading } from './render';
import { EpisodeSelectBreadcrumbs } from './SeasonSelectComponent';
import { MLink } from './utils';
import { DownloadCall } from './DownloadComponent';
import { Torrents } from './streaming';
import { DisplayError } from './IndexComponent';

type MapType = [string, ITorrent[]][];

function DownloadAllComponent() {
  const { tmdb_id, season: season_s } = useParams<{
    tmdb_id: string;
    season: string;
  }>();
  const season = parseInt(season_s!);

  const { data: torrents } = useSWR<Torrents>('torrents');
  const { data, isValidating, error } = useSWR<
    {
      packs: ITorrent[];
      complete: MapType;
      incomplete: MapType;
    },
    Error
  >(`select/${tmdb_id}/season/${season}/download_all`);

  return (
    <div>
      {error && <DisplayError error={error} />}
      <EpisodeSelectBreadcrumbs tmdb_id={tmdb_id!} season={season} />
      <Loading loading={isValidating} />
      <div>
        <h3>Packs</h3>
        <ul>
          {data &&
            data.packs &&
            data.packs.map((t) => (
              <li key={t.download}>
                <DisplayTorrent
                  torrents={torrents}
                  torrent={t}
                  tmdb_id={tmdb_id!}
                  season={season}
                />
              </li>
            ))}
        </ul>
      </div>
      <Individual
        label="Complete Sets"
        items={data && data.complete}
        season={season}
        tmdb_id={tmdb_id!}
        torrents={torrents}
      />
      <Individual
        label="Incomplete Sets"
        items={data && data.incomplete}
        season={season}
        tmdb_id={tmdb_id!}
        torrents={torrents}
      />
    </div>
  );
}

function download_all(tmdb_id: number, torrents: ITorrent[]) {
  const downloads: DownloadCall[] = torrents.map((t) => ({
    tmdb_id,
    magnet: t.download,
    season: t.episode_info?.seasonnum,
    episode: t.episode_info?.epnum,
  }));

  return { to: '/download', state: { downloads } };
}

function Individual(props: {
  label: string;
  items?: MapType;
  season: number;
  tmdb_id: string;
  torrents?: Torrents;
}) {
  return (
    <div>
      <h3>{props.label}</h3>
      <ul>
        {props.items &&
          props.items.map(([name, torrents]) => (
            <div key={name}>
              <h4>
                <MLink {...download_all(parseInt(props.tmdb_id), torrents)}>
                  {name}
                </MLink>
              </h4>
              <ul>
                {torrents.map((t) => (
                  <li key={t.download}>
                    <DisplayTorrent
                      torrents={props.torrents}
                      torrent={t}
                      tmdb_id={props.tmdb_id}
                      season={props.season}
                      episode={t.episode_info?.epnum}
                    />
                  </li>
                ))}
              </ul>
            </div>
          ))}
      </ul>
      {props.items && !props.items.length && <p>No results found</p>}
    </div>
  );
}

export { DownloadAllComponent };
