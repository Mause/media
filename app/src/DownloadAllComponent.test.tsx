import { MemoryRouter, Route, Routes } from 'react-router-dom';
import { http, HttpResponse } from 'msw';

import { renderWithSWR, waitForRequests } from './test.utils';
import { DownloadAllComponent } from './DownloadAllComponent';
import type { ITorrent } from './OptionsComponent';
import { server } from './msw';

test('DownloadAllComponent', async () => {
  server.use(
    http.get('/api/select/1/season/1/download_all', () =>
      HttpResponse.json({
        packs: [
          {
            source: 'horriblesubs',
            category: 'Movie',
            download: 'magnets://?xt=hello',
            episode_info: { seasonnum: 1, epnum: 1 },
            seeders: 5,
            title: 'Hello World',
          },
        ] satisfies ITorrent[],
        incomplete: [],
        complete: [],
      }),
    ),
  );
  server.use(http.get('/api/tv/1', () => HttpResponse.json({})));
  server.use(http.get('/api/torrents', () => HttpResponse.json({})));

  const { container } = renderWithSWR(
    <MemoryRouter initialEntries={['/select/1/season/1/download_all']}>
      <Routes>
        <Route
          path="/select/:tmdb_id/season/:season/download_all"
          Component={DownloadAllComponent}
        />
      </Routes>
    </MemoryRouter>,
  );

  expect(container).toMatchSnapshot();

  await waitForRequests();

  expect(container).toMatchSnapshot();
});
