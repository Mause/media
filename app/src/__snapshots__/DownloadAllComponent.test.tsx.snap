// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DownloadAllComponent 1`] = `
<div>
  <div>
    <nav
      aria-label="breadcrumb"
      class="MuiTypography-root MuiTypography-body1 MuiBreadcrumbs-root css-5cvx94-MuiTypography-root-MuiBreadcrumbs-root"
    >
      <ol
        class="MuiBreadcrumbs-ol css-1uwp4ue-MuiBreadcrumbs-ol"
      >
        <li
          class="MuiBreadcrumbs-li"
        >
          <nav
            class="MuiTypography-root MuiTypography-body1 MuiBreadcrumbs-root css-5cvx94-MuiTypography-root-MuiBreadcrumbs-root"
          >
            <ol
              class="MuiBreadcrumbs-ol css-1uwp4ue-MuiBreadcrumbs-ol"
            >
              <li
                class="MuiBreadcrumbs-li"
              >
                <a
                  class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineHover css-nv1n36-MuiTypography-root-MuiLink-root"
                  href="/"
                >
                  Home
                </a>
              </li>
              <li
                aria-hidden="true"
                class="MuiBreadcrumbs-separator css-jhr02i-MuiBreadcrumbs-separator"
              >
                /
              </li>
              <li
                class="MuiBreadcrumbs-li"
              >
                <p
                  class="MuiTypography-root MuiTypography-body1 css-rizt0-MuiTypography-root"
                >
                  Search Results
                </p>
              </li>
            </ol>
          </nav>
        </li>
        <li
          aria-hidden="true"
          class="MuiBreadcrumbs-separator css-jhr02i-MuiBreadcrumbs-separator"
        >
          /
        </li>
        <li
          class="MuiBreadcrumbs-li"
        >
          <a
            class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineHover css-nv1n36-MuiTypography-root-MuiLink-root"
            href="/select/1/season"
          />
        </li>
        <li
          aria-hidden="true"
          class="MuiBreadcrumbs-separator css-jhr02i-MuiBreadcrumbs-separator"
        >
          /
        </li>
        <li
          class="MuiBreadcrumbs-li"
        >
          <p
            class="MuiTypography-root MuiTypography-body1 css-1bgbmvw-MuiTypography-root"
            data-testid="title"
          >
            Season 
            1
          </p>
        </li>
      </ol>
    </nav>
    <svg
      aria-hidden="true"
      class="svg-inline--fa fa-spinner fa-w-16 fa-spin fa-xs "
      data-icon="spinner"
      data-prefix="fas"
      focusable="false"
      role="img"
      viewBox="0 0 512 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M304 48c0 26.51-21.49 48-48 48s-48-21.49-48-48 21.49-48 48-48 48 21.49 48 48zm-48 368c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.49-48-48-48zm208-208c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.49-48-48-48zM96 256c0-26.51-21.49-48-48-48S0 229.49 0 256s21.49 48 48 48 48-21.49 48-48zm12.922 99.078c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48c0-26.509-21.491-48-48-48zm294.156 0c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48c0-26.509-21.49-48-48-48zM108.922 60.922c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.491-48-48-48z"
        fill="currentColor"
      />
    </svg>
    <div>
      <h3>
        Packs
      </h3>
      <ul />
    </div>
    <div>
      <h3>
        Complete Sets
      </h3>
      <ul />
    </div>
    <div>
      <h3>
        Incomplete Sets
      </h3>
      <ul />
    </div>
  </div>
</div>
`;

exports[`DownloadAllComponent 2`] = `
<div>
  <div>
    <nav
      aria-label="breadcrumb"
      class="MuiTypography-root MuiTypography-body1 MuiBreadcrumbs-root css-5cvx94-MuiTypography-root-MuiBreadcrumbs-root"
    >
      <ol
        class="MuiBreadcrumbs-ol css-1uwp4ue-MuiBreadcrumbs-ol"
      >
        <li
          class="MuiBreadcrumbs-li"
        >
          <nav
            class="MuiTypography-root MuiTypography-body1 MuiBreadcrumbs-root css-5cvx94-MuiTypography-root-MuiBreadcrumbs-root"
          >
            <ol
              class="MuiBreadcrumbs-ol css-1uwp4ue-MuiBreadcrumbs-ol"
            >
              <li
                class="MuiBreadcrumbs-li"
              >
                <a
                  class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineHover css-nv1n36-MuiTypography-root-MuiLink-root"
                  href="/"
                >
                  Home
                </a>
              </li>
              <li
                aria-hidden="true"
                class="MuiBreadcrumbs-separator css-jhr02i-MuiBreadcrumbs-separator"
              >
                /
              </li>
              <li
                class="MuiBreadcrumbs-li"
              >
                <p
                  class="MuiTypography-root MuiTypography-body1 css-rizt0-MuiTypography-root"
                >
                  Search Results
                </p>
              </li>
            </ol>
          </nav>
        </li>
        <li
          aria-hidden="true"
          class="MuiBreadcrumbs-separator css-jhr02i-MuiBreadcrumbs-separator"
        >
          /
        </li>
        <li
          class="MuiBreadcrumbs-li"
        >
          <a
            class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineHover css-nv1n36-MuiTypography-root-MuiLink-root"
            href="/select/1/season"
          />
        </li>
        <li
          aria-hidden="true"
          class="MuiBreadcrumbs-separator css-jhr02i-MuiBreadcrumbs-separator"
        >
          /
        </li>
        <li
          class="MuiBreadcrumbs-li"
        >
          <p
            class="MuiTypography-root MuiTypography-body1 css-1bgbmvw-MuiTypography-root"
            data-testid="title"
          >
            Season 
            1
          </p>
        </li>
      </ol>
    </nav>
    <div>
      <h3>
        Packs
      </h3>
      <ul>
        <li>
          <span>
            <strong
              title="horriblesubs"
            >
              H
            </strong>
             
            <a
              class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineHover css-73o1y3-MuiTypography-root-MuiLink-root"
              href="/download"
            >
              Hello World
            </a>
             
            <small>
              5
            </small>
             
          </span>
        </li>
      </ul>
    </div>
    <div>
      <h3>
        Complete Sets
      </h3>
      <ul />
      <p>
        No results found
      </p>
    </div>
    <div>
      <h3>
        Incomplete Sets
      </h3>
      <ul />
      <p>
        No results found
      </p>
    </div>
  </div>
</div>
`;
